### PR TITLE
Add Sensor-Rate logging to BatchSampler

### PR DESCRIPTION
**Pull request by peterbarker**
Typically our sensors down-sample the data coming from the sensor before feeding it into the frontend for further processing.  This feeding is usually done at 1kHz, and that was the sample rate used by the existing BatchSampling implementation.

This PR adds the facility for a backend to provide data to the BatchSampler at its sensor-rate.  Currently this is only implemented on the InvenSense IMUs, but this allows them to log data at their full sensor rate - 8kHz gyros and 4kHz accelerometers.

Having this data will allow builders to resolve vibration issues more readily, in addition to making cool-looking graphs.

This is a pair of commits; tridge's initial hackery and my cleanups and generalisations.

### CODE REVIEW

**Comment by peterbarker** [comment]
https://github.com/ArduPilot/ardupilot_wiki/pull/1143 <- wiki updates for this PR

File: libraries/AP_InertialSensor/AP_InertialSensor.h [hunk 1]
```diff
@@ -294,17 +294,37 @@ class AP_InertialSensor : AP_AccelCal_Client
...
         static const struct AP_Param::GroupInfo var_info[];
 
         // Parameters
         AP_Int16 _required_count;
         AP_Int8 _sensor_mask;
+        AP_Int8 _batch_options_mask;
+
+        // push blocks to DataFlash at regular intervals.  each
+        // message is ~ 108 bytes in size, so we use about 1kB/s of
+        // logging bandwidth with a 100ms interval.  If we are taking
```
Comment (OXINARF) [inline libraries/AP_InertialSensor/AP_InertialSensor.h]: Comment is now wrong.

**Comment by OXINARF** [comment]
Forgot one thing: _sensorrate_ looks a bit bad; in most places, it would be perfectly fine to be _sensor_rate_ (or _sensor rate_ in comments)

**Comment by peterbarker** [comment]
Comment fixed, and sensorrate changed.

tridge has suggested we remove the fast logging at some stage; it does nasty things to performance.

So currently we have:
 IMU (log bitmask)
FASTIMU (log bitmask)
BatchSampler (1kHz)
BatchSampler (sensor-rate)

*and* they interact, so you can't do BatchSampler while doing FASTIMU.  Sadly, people do seem to use FASTIMU...

**Comment by amilcarlucas** [comment]
Can we document the invalid combinations on the wiki?

**Comment by amilcarlucas** [comment]
Can we document the invalid combinations on the wiki?

**Comment by peterbarker** [comment]
On Thu, 5 Apr 2018, Amilcar Lucas wrote:

> Can we document the invalid combinations on the wiki?

Yes, but whereever we put it it won't be found.  Better to eliminate the 
conflicts, I think.  We can do that when people start to use the 
batchsampled data as a matter of course in place of fastimu

**Pull request merged**